Fix Dependency File Detection for npm, pnpm, yarn, and bun

kbukum1 opened the pull request:
### What are you trying to accomplish?

This PR fixes an issue with detecting dependency files in workspaces when file names include parent directories or relative paths (e.g., `./../pnpm-lock.yaml`). The previous logic relied solely on exact name matches (`==`), which caused package manager detection to fail when no lock file was found. 

The updated logic uses both `==` and `end_with?` checks to improve robustness and ensure compatibility with various file naming conventions.

### Anything you want to highlight for special attention from reviewers?

- Changes have been applied to methods that handle lockfiles and configuration files for `npm`, `pnpm`, `yarn`, and `bun`.
- This fix addresses edge cases where dependency files are nested or include parent directory references in their paths.
- Reviewers should check if all relevant file types are covered and ensure the changes do not introduce regressions.

### How will you know you've accomplished your goal?

- Dependency files are correctly identified in all scenarios, including nested or prefixed paths.
- Tests verify correct behavior for each package manager (`npm`, `pnpm`, `yarn`, and `bun`).
- Package manager detection works without errors caused by missing or misidentified files.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

abdulapopoola commented on npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:
```diff
@@ -143,56 +143,56 @@ def package_json
       def shrinkwrap
         @shrinkwrap ||= T.let(dependency_files.find do |f|
-          f.name == NpmPackageManager::SHRINKWRAP_LOCKFILE_NAME
+          f.name.end_with?(NpmPackageManager::SHRINKWRAP_LOCKFILE_NAME)
```
do we want to check for case insensitivity?

kbukum1 replied:
I think we are good for case sensitive. Let me check but file_fetcher is working fine by using case sensitive namings.

kbukum1 replied:
Maybe we can. I can see in file fetcher we are doing following to set name. It may not be small letters.

```
        DependencyFile.new(
          name: Pathname.new(filename).cleanpath.to_path,
          directory: directory,
          type: type,
          content: content,
          symlink_target: symlink_target,
          support_file: in_submodule?(path)
        )
   ```

kbukum1 replied:
Normally we only have problem with lock files where we have following if we can't find file from the directory. 

```
      sig { params(filename: String).returns(T.nilable(DependencyFile)) }
      def fetch_file_from_parent_directories(filename)
        (1..directory.split("/").count).each do |i|
          file = fetch_file_with_support(("../" * i) + filename)
          return file if file
        end
        nil
      end
```

kbukum1 replied:
Let me try.

kbukum1 commented on npm_and_yarn/helpers/lib/npm/vulnerability-auditor.js:
```diff
@@ -269,23 +269,23 @@ const maybeReadFile = file => {
+  return raw.replace(/\r\n/g, '\n').split(delim)
+    .filter(section => section.trim())
+    .map(section => section.trimStart() + delim)
 }
```
The IDE's automatic formatter is adjusting the code indentation to a tabSize of 2 instead of 4, which deviates from our standard indentation style.

kbukum1 commented on go_modules/spec/fixtures/projects/module_major_version_mismatch_v1/go.sum:
```diff
@@ -195,7 +195,6 @@ github.com/xiang90/probing v0.0.0-20190116061207-43a291ad63a2/go.mod h1:UETIi67q
 github.com/xordataexchange/crypt v0.0.3-0.20170626215501-b2862e3d0a77/go.mod h1:aYKd//L2LvnjZzWKhF00oedf4jCCReLcmhLdhm1A27Q=
 github.com/yuin/goldmark v1.1.27/go.mod h1:3hX8gzYuyVAZsxl0MRgGTJEmQBFcNTphYh9decYSb74=
 go.etcd.io/bbolt v1.3.2/go.mod h1:IbVyRI1SCnLcuJnV2u8VeU0CEYM7e686BmAb1XKL+uU=
-go.opencensus.io v0.21.0/go.mod h1:mSImk1erAIZhrmZN+AvHh14ztQfjbGwt4TtuofqLduU=
```
Just removed the library because it was causing the following error in go specs

```
       expected "go: github.com/jenkins-x/jx-api@v0.0.24 requires\n\tk8s.io/client-go@v0.16.5 requires\n\tcloud.google.com/go@v0.38.0 requires\n\tgo.opencensus.io@v0.21.0: unrecognized import path \"go.opencensus.io\": https fetch: Get \"https://go.opencensus.io/?go-get=1\": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2025-01-22T00:44:17Z is after 2025-01-21T03:43:04Z" to include "go.mod has post-v1 module path"
       Diff:
       @@ -1,4 +1,7 @@
       -go.mod has post-v1 module path
       +go: github.com/jenkins-x/jx-api@v0.0.24 requires
       +	k8s.io/client-go@v0.16.5 requires
       +	cloud.google.com/go@v0.38.0 requires
       +	go.opencensus.io@v0.21.0: unrecognized import path "go.opencensus.io": https fetch: Get "https://go.opencensus.io/?go-get=1": tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2025-01-22T00:44:17Z is after 2025-01-21T03:43:04Z
     # ./spec/dependabot/go_modules/file_updater/go_mod_updater_spec.rb:558:in `block (4 levels) in <top (required)>'
     # /home/dependabot/common/spec/spec_helper.rb:66:in `block (2 levels) in <top (required)>'
     # /home/dependabot/dependabot-updater/vendor/ruby/3.3.0/gems/webmock-3.24.0/lib/webmock/rspec.rb:39:in `block (2 levels) in <top (required)>'
 
```

kbukum1 commented:
Since the test focuses on path validation for `go_modules`, verifying the `v1` path against the `v2` path is sufficient. Adding a similar check for `v0` is redundant and does not provide additional value to the test. Therefore, I have updated the existing spec to focus on the `v1` path check and removed the redundant spec for the `v0` path check.
